### PR TITLE
Implemented PolicyRegistry

### PR DESCRIPTION
**Pull request by rpartsey**
## Motivation and Context

Often we want to swap policy architecture but use the same trainer.
See issue #404.

## How Has This Been Tested
Tests and pre-commit hooks passed successfully.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: habitat_baselines/common/policy_registry.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+from typing import Optional
+
+from habitat.core.registry import Registry
+
+
+class PolicyRegistry(Registry):
+    @classmethod
+    def register_policy(cls, to_register=None, *, name: Optional[str] = None):
+        r"""Register a RL policy to registry with key 'name'.
```
Comment (Skylion007) [inline habitat_baselines/common/policy_registry.py:27]: The doc strings are unfortunately not in the right format for our documentation system. Please see the `:param:` syntax we use elsewhere.

File: habitat_baselines/common/policy_registry.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        """
+        from habitat_baselines.rl.ppo.policy import Policy
+
+        return cls._register_impl(
+            "policy", to_register, name, assert_type=Policy
+        )
+
+    @classmethod
+    def get_policy(cls, name):
```
Comment (Skylion007) [inline habitat_baselines/common/policy_registry.py:41]:
```suggestion
    def get_policy(cls, name : str):
```

**Comment by rpartsey** [comment]
**To discuss:** How to properly handle instantiating policies dynamically

1. Come up with a standard interface for policies?
2. Use similar approach as `make_task`?
3. Other ...

File: habitat_baselines/common/policy_registry.py (hunk 2 above)
Comment (Skylion007) [inline habitat_baselines/common/policy_registry.py:41]: Also should have a doc string.

File: habitat_baselines/common/policy_registry.py (hunk 1 above)
Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py:27]: Thanks!
Will fix that
I formatted docstring the same way as in the [BaselineRegistry](https://github.com/facebookresearch/habitat-api/blob/7f2121e1a31d304b5e897c8a97929de44b4f02aa/habitat_baselines/common/baseline_registry.py#L29).

Comment (Skylion007) [reply habitat_baselines/common/policy_registry.py:27]: Actually, you might be right that we may not build docs for habitat_baselines, or habitat-api. Perhaps we should consider doing so if we don't.

Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py]: @Skylion007  Updated!

File: habitat_baselines/common/policy_registry.py (hunk 2 above)
Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py]: Updated
**Note:** other `get_somethint` methods in the Registry and BaselineRegistry don't have type annotations.

Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py]: Updated
**Note:** other `get_somethint` methods in the Registry and BaselineRegistry don't have docstrings.

Comment (Skylion007) [reply habitat_baselines/common/policy_registry.py]: Can't hurt 👍

File: habitat_baselines/common/policy_registry.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+-   Register a policy: ``@policy_registry.register_policy``
+"""
+
+from typing import Optional
+
+from habitat.core.registry import Registry
+
+
+class PolicyRegistry(Registry):
```
Comment (erikwijmans) [inline habitat_baselines/common/policy_registry.py:26]: This doesn't need to be a separate class.  The two methods can be added to the existing BaselineRegistry.  The `"policy"` arg used bellow acts as an additional key so it won't conflict with anything else there.

Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py]: Done!

**Comment by rpartsey** [comment]
I'd also suggest creating a new issue to standardize the docstrings, type annotations, and methods definition order of `Registry` and `BaselineRegistry` classes.

What do you think?

**Comment by Skylion007** [comment]
Sounds good!

File: habitat_baselines/common/policy_registry.py (hunk 1 above)
Comment (rpartsey) [reply habitat_baselines/common/policy_registry.py]: @Skylion007 would you mind resolving this change request?
I've already updated code according to your comments

File: habitat_baselines/config/test/ppo_imagenav_test.yaml [hunk 4]
```diff
@@ -13,6 +13,9 @@ LOG_INTERVAL: 100
 CHECKPOINT_INTERVAL: 1
 
 RL:
+  POLICY:
+    name: "PointNavBaselinePolicy"
```
Comment (mathfac) [inline habitat_baselines/config/test/ppo_imagenav_test.yaml:17]: Please, add default "POLICY.name" value to the default config [here](https://github.com/facebookresearch/habitat-api/blob/master/habitat_baselines/config/default.py#L51). Then we can drop adding policy in part of configs. I would set default to `PointNavBaselinePolicy` as PPO trainer is default one.

File: habitat_baselines/rl/ppo/policy.py [hunk 5]
```diff
@@ -88,8 +89,11 @@ def forward(self, x):
         return self.fc(x)
 
 
+@baseline_registry.register_policy
 class PointNavBaselinePolicy(Policy):
-    def __init__(self, observation_space, action_space, hidden_size=512):
+    def __init__(
+        self, observation_space, action_space, hidden_size=512, **kwargs
```
Comment (mathfac) [inline habitat_baselines/rl/ppo/policy.py:95]: Maybe, worth to add ` **kwargs` to [PointNavResNetPolicy](https://github.com/facebookresearch/habitat-api/blob/master/habitat_baselines/rl/ddppo/policy/resnet_policy.py#L34).

File: habitat_baselines/rl/ppo/ppo_trainer.py [hunk 6]
```diff
@@ -59,10 +59,17 @@ def _setup_actor_critic_agent(self, ppo_cfg: Config) -> None:
         """
         logger.add_filehandler(self.config.LOG_FILE)
 
-        self.actor_critic = PointNavBaselinePolicy(
+        policy = baseline_registry.get_policy(self.config.RL.POLICY.name)
+        self.actor_critic = policy(
             observation_space=self.envs.observation_spaces[0],
             action_space=self.envs.action_spaces[0],
             hidden_size=ppo_cfg.hidden_size,
+            rnn_type=self.config.RL.DDPPO.rnn_type,
```
Comment (mathfac) [inline habitat_baselines/rl/ppo/ppo_trainer.py:67]: I had a second look on this design choice, accessing `RL.DDPPO` looks strange here, as it's different baseline. Let's remove new unrelated arguments and let's the run fails if wrong Policy was specified.

File: habitat_baselines/config/test/ppo_imagenav_test.yaml (hunk 4 above)
Comment (rpartsey) [reply habitat_baselines/config/test/ppo_imagenav_test.yaml]: Done!

File: habitat_baselines/rl/ppo/policy.py (hunk 5 above)
Comment (rpartsey) [reply habitat_baselines/rl/ppo/policy.py:95]: Done!

File: habitat_baselines/rl/ppo/ppo_trainer.py (hunk 6 above)
Comment (rpartsey) [reply habitat_baselines/rl/ppo/ppo_trainer.py]: Done!

File: habitat_baselines/common/baseline_registry.py [hunk 7]
```diff
@@ -61,5 +64,42 @@ def register_env(cls, to_register=None, *, name: Optional[str] = None):
     def get_env(cls, name):
         return cls._get_impl("env", name)
 
+    @classmethod
+    def register_policy(cls, to_register=None, *, name: Optional[str] = None):
+        r"""Register a RL policy to registry with :p:`name`.
```
Comment (mathfac) [inline habitat_baselines/common/baseline_registry.py:69]:
```suggestion
        r"""Register a RL policy with :p:`name`.
```

**Pull request merged**